File: src/main/java/com/anod/appwatcher/recyclerview/RecyclerViewCursorAdapter.java
package com.anod.appwatcher.recyclerview;

import android.content.Context;
import android.database.Cursor;
import android.database.CursorWrapper;
import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * @author alex
 * @date 2015-06-20
 */
public abstract class RecyclerViewCursorAdapter<VH extends RecyclerView.ViewHolder, CR extends CursorWrapper> extends RecyclerView.Adapter<VH> {
    private boolean mDataValid;
    private final Context mContext;
    private final int mResource;
    private CR mCursor;

    public RecyclerViewCursorAdapter(Context context, @LayoutRes int itemResource) {
        mContext = context;
        mResource = itemResource;
    }

    protected abstract VH onCreateViewHolder(View itemView);
    protected abstract void onBindViewHolder(VH holder, int position, CR cursor);

    @Override
    public VH onCreateViewHolder(ViewGroup parent, int viewType) {
        return onCreateViewHolder(createItemView(parent, viewType));
    }

    @Override
    public void onBindViewHolder(VH holder, int position) {
        if (!mCursor.moveToPosition(position)) {
            throw new IllegalStateException("couldn't move cursor to position " + position);
        }
        onBindViewHolder(holder, position, mCursor);
    }

    @Override
    public int getItemCount() {
        if (mDataValid && mCursor != null) {
            return mCursor.getCount();
        } else {
            return 0;
        }
    }

    @Override
    public int getItemViewType(int position) {
        return 1;
    }

    protected View createItemView(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(mContext).inflate(mResource, parent, false);
        v.setClickable(true);
        v.setFocusable(true);
        return v;
    }

    public void swapData(CR newCursor) {
        if (newCursor == mCursor) {
            return;
        }
        mCursor = newCursor;
        if (newCursor != null) {
            mDataValid = true;
            // notify the observers about the new cursor
            notifyDataSetChanged();
        } else {
            mDataValid = false;
            // notify the observers about the lack of a data set
            notifyDataSetChanged();
        }
    }
}

File: src/main/java/com/anod/appwatcher/framework/RecyclerViewCursorAdapter.kt
package com.anod.appwatcher.framework

import android.content.Context
import android.database.CursorWrapper
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

/**
 * @author alex
 * *
 * @date 2015-06-20
 */
abstract class RecyclerViewCursorAdapter<VH : RecyclerView.ViewHolder, in CR : CursorWrapper>(
        private val context: Context,
        @param:LayoutRes private val resource: Int) : RecyclerView.Adapter<VH>() {

    private var mDataValid: Boolean = false
    private var mCursor: CR? = null

    protected abstract fun onCreateViewHolder(itemView: View): VH
    protected abstract fun onBindViewHolder(holder: VH, position: Int, cursor: CR)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        return onCreateViewHolder(createItemView(parent, viewType))
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        if (mCursor == null || !mCursor!!.moveToPosition(position)) {
            throw IllegalStateException("couldn't move cursor to position " + position)
        }
        onBindViewHolder(holder, position, mCursor!!)
    }

    override fun getItemCount(): Int {
        if (mDataValid && mCursor != null) {
            return mCursor!!.count
        } else {
            return 0
        }
    }

    override fun getItemViewType(position: Int): Int {
        return 1
    }

    protected fun createItemView(parent: ViewGroup, viewType: Int): View {
        val v = LayoutInflater.from(context).inflate(resource, parent, false)
        v.isClickable = true
        v.isFocusable = true
        return v
    }

    open fun swapData(newCursor: CR?) {
        if (newCursor === mCursor) {
            return
        }
        mCursor = newCursor
        if (newCursor != null) {
            mDataValid = true
            // notify the observers about the new cursor
            notifyDataSetChanged()
        } else {
            mDataValid = false
            // notify the observers about the lack of a data set
            notifyDataSetChanged()
        }
    }
}
